### PR TITLE
Add support for partial ARNs for EventSourceMappings and fix AWS::Lambda::Version state check

### PR DESCRIPTION
**Pull request by dominikschubert**
## Motivation

Investigation started due to a bug report that turned out to not actually be a bug (Select/ Join intrinsic functions).

At its core it was trying to create an `AWS::Lambda::EventSourceMapping` resource targeting a partial ARN like `<function_name>:live` which we didn't properly handle in the underlying `CreateEventSourceMapping` handler in the lambda API provider.

## Changes

- Adds a test which creates event source mappings with different structures for the `FunctionName` parameter.
- Add support for partial ARNs and other edge cases in `CreateEventSourceMapping`
- Fix `create` handler for `AWS::Lambda::Version` to properly wait for the function version to be active (cc @Morijarti ). This was causing flakes and is a regression introduced in the migration from the old to the new CFn providers.

### CODE REVIEW

File: localstack/services/lambda_/provider.py [hunk 1]
```diff
@@ -1665,21 +1665,36 @@ def create_event_source_mapping(
...
-        if api_utils.FULL_FN_ARN_PATTERN.match(function_name):
-            fn_arn = function_name
-            function_name = api_utils.get_function_name(function_name, context)
-        else:
-            fn_arn = api_utils.unqualified_lambda_arn(
-                function_name, context.account_id, context.region
-            )
+        request_function_name = request["FunctionName"]
+        # can be either a partial arn or a full arn for the version/alias
+        function_name, qualifier, account, region = function_locators_from_arn(
```
Comment (joe4dev) [inline localstack/services/lambda_/provider.py:1670]: nice migration to use this most powerful util 👍

File: localstack/services/lambda_/resource_providers/aws_lambda_version.py [hunk 2]
```diff
@@ -58,19 +58,40 @@ def create(
...
-
-        model["Version"] = response["Version"]
-        model["Id"] = response["FunctionArn"]
-
-        return ProgressEvent(
-            status=OperationStatus.SUCCESS,
-            resource_model=model,
-            custom_context=request.custom_context,
-        )
+        if not ctx.get(REPEATED_INVOCATION):
```
Comment (joe4dev) [inline localstack/services/lambda_/resource_providers/aws_lambda_version.py:65]: 💡  interesting pattern, probably gonna be required in a couple of places

File: localstack/services/lambda_/provider.py [hunk 3]
```diff
@@ -1665,21 +1665,36 @@ def create_event_source_mapping(
...
+            request_function_name
+        )
+        account = account or context.account_id
```
Comment (joe4dev) [inline localstack/services/lambda_/provider.py:1673]: FYI: we have not tested cross-account behavior for create operations in the Lambda core API yet

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 4]
```diff
@@ -4324,10 +4327,65 @@ def check_esm_active():
...
+
+        _create_esm("name_only", function_name)
+        _create_esm("partial_arn_latest", f"{function_name}:$LATEST")
+        _create_esm("partial_arn_version", f"{function_name}:{v1['Version']}")
+        _create_esm("partial_arn_alias", f"{function_name}:{alias['Name']}")
+        _create_esm("full_arn_latest", fn["Configuration"]["FunctionArn"])
+        _create_esm("full_arn_version", v1["FunctionArn"])
+        _create_esm("full_arn_alias", alias["AliasArn"])
+
+    # def test_multiple_esm_conflict(self):
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:4381]: is this a WIP test case?

**Review by joe4dev — APPROVED** [review]
LGTM 👍 

There is some WIP code to cleanup.

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 4 above)
Comment (dominikschubert) [reply tests/aws/services/lambda_/test_lambda_api.py:4381]: more like a TODO I guess if anyone is picking up general work on EventSourceMappings in the future :eyes: 

Basically this occurs when you try to create an event source mapping for a combination of function + target ARN that already exists.

**Comment by joe4dev** [comment]
The failing test looks like a flake `FAILED tests/aws/services/lambda_/test_lambda_common.py::TestLambdaRuntimesCommon::test_uncaught_exception_invoke[provided]`
The image `provided` (and many others) was recently updated but I cannot reproduce the failing test locally.
Merging ...

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 4 above)
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py]: okay, already merged 😎 
I can sneak in a TODO in a later PR to add that clarification to the exception tests ...

**Pull request merged**